vllm/core/scheduler.py

```diff
@@ -882,6 +922,17 @@ def _schedule_prefills(
             assert len(waiting_seqs) == 1, (
                 "Waiting sequence group should have only one prompt "
                 "sequence.")
+
+            is_big = self._is_big_seq_group(seq_group)
+            if is_big and self.big_prefill_requests >= self.max_big_requests \
+                and self.num_prefill_slots > 1:
+                # When we have more than one prefill slot, we limit the number
+                # of big requests to avoid filling all of our slots with partial
+                # prefills of big prompt sequences.
+                leftover_waiting_sequences.appendleft(seq_group)
+                waiting_queue.popleft()
+                continue
+
             num_new_tokens = self._get_num_new_tokens(seq_group,
                                                       SequenceStatus.WAITING,
                                                       enable_chunking, budget)
```

Comment:
In the case where we only have 1 `prefill-slot` and we send in 2 concurrent requests, one big and one small, when the scheduler is trying to `schedule_prefill`, it fills up the token budget with the first large request. When it loops through the waiting requests and tries to get `num_new_tokens` for the 2nd request, it gets 0 since budget is 0. Can we break out earlier instead of breaking out way down on line 990?

Reply:
Yeah that sounds reasonable. Maybe at the very end of the loop after we schedule a request, we can check if budget == 0 and `break` instead of looping to the next request?

Could be ⚡⚡⚡ 🚀

Reply:
I think in cases where you have 1 prefill slot and a large request running, most of the time it'll use up the budget inside of `_schedule_running` and hit the check at the top of `_schedule_prefills` and not loop at all, but this optimization would still be useful. In cases with constant load of small requests, you could potentially save one loop iteration most times here